packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart

```diff
@@ -312,6 +333,92 @@ class DebuggerController extends DisposableController
     appendStdio(text);
   }
 
+  Future<List<ScriptRef>> _retrieveSortScripts(IsolateRef ref) async {
+    final scriptList = await _service.getScripts(isolateRef.id);
+    // We filter out non-unique ScriptRefs here (dart-lang/sdk/issues/41661).
+    final scriptRefs = Set.of(scriptList.scripts).toList();
+    scriptRefs.sort((a, b) {
+      // We sort uppercase so that items like dart:foo sort before items like
+      // dart:_foo.
+      return a.uri.toUpperCase().compareTo(b.uri.toUpperCase());
+    });
+    return scriptRefs;
+  }
+
+  void _updateAfterIsolateReload(Event reloadEvent) async {
+    // Generally this has the value 'success'; we update our data in any case.
+    // ignore: unused_local_variable
+    final status = reloadEvent.status;
+
+    // Refresh the list of scripts.
+    final scriptRefs = await _retrieveSortScripts(isolateRef);
+    for (var scriptRef in scriptRefs) {
+      _uriToScriptMap[scriptRef.uri] = scriptRef;
+    }
+
+    final removedScripts =
+        Set.of(_sortedScripts.value).difference(Set.of(scriptRefs));
+    final addedScripts =
+        Set.of(scriptRefs).difference(Set.of(_sortedScripts.value));
+
+    _sortedScripts.value = scriptRefs;
+
+    // TODO(devoncarew): Show a message in the logging view.
+
+    // Show a toast.
+    final count = removedScripts.length + addedScripts.length;
+    messageBus.addEvent(BusEvent('toast',
+        data: '${nf.format(count)} ${pluralize('script', count)} updated.'));
+
+    // Update breakpoints.
+    _updateBreakpointsAfterReload(removedScripts, addedScripts);
+
+    // Redirect the current editor screen if necessary.
+    if (removedScripts.contains(currentScriptRef.value)) {
+      final uri = currentScriptRef.value.uri;
+      final newScriptRef = addedScripts
+          .firstWhere((script) => script.uri == uri, orElse: () => null);
+
+      if (newScriptRef != null) {
+        // Pre-prime the script info.
+        // ignore: unawaited_futures
+        getScript(newScriptRef).then((script) {
+          showScriptLocation(ScriptLocation(newScriptRef));
+        });
+      }
+    }
+
+    // TODO(devoncarew): Invalidate the list of classes?
+  }
+
+  void _updateBreakpointsAfterReload(
+      Set<ScriptRef> removedScripts, Set<ScriptRef> addedScripts) {
+    // TODO(devoncarew): We need to coordinate this with other debugger clients
+    // as well as pause before re-setting the breakpoints.
+
+    final breakpointsToRemove = <BreakpointAndSourcePosition>[];
+
+    for (final scriptRef in removedScripts) {
+      for (final bp in breakpointsWithLocation.value) {
+        if (bp.scriptRef == scriptRef) {
+          breakpointsToRemove.add(bp);
+        }
+      }
+    }
+
+    for (final bp in breakpointsToRemove) {
+      removeBreakpoint(bp.breakpoint);
+    }
+
+    for (final scriptRef in addedScripts) {
+      for (final bp in breakpointsToRemove) {
```

Comment:
I'm confused why this is looping through breakpoints to remove and then adding them. Is this intentional? and if so can you add a comment?

Reply:
I'll add some documentation - we're removing breakpoints from scripts where newer versions of those scripts are available.